### DISCUSSION TITLE
Middleware that checks result and rolls back transaction

### DISCUSSION THREAD

**Discussion by mdesousa** [prisma/prisma#9687] [Q&A] [answered]
Hi, I am looking for a way to implement middleware that can validate the result of an update and potentially rollback the update if validation fails. This would be very useful for validation of business rules when partial updates are done.
An example would be updating the `start_date` column in a table, and checking that `start_date <= `end_date`.

I am trying to write the middleware as shown below. However, it appears that the transaction has completed by the time that the error is thrown. Is there a way to rollback the transaction?

```ts
prisma.$use(async (params, next) => {
  const promise = next(params);
  if (params.action !== 'update') return promise;
  const result = await promise;
  if (result.start_date <= result.end_date) return result;
  throw Error('start_date should happen before end_date');
});
```

**Comment by ryands17** [comment]
@mdesousa 👋 

This is not possible using middleware, I would suggest using [Interactive Transactions](https://www.prisma.io/docs/concepts/components/prisma-client/transactions#interactive-transactions-in-preview) for this use case.

**Reply by mdesousa** [reply]
Thanks cousin @ryands17 😆 
Actually the interactive transactions look very useful, i'll check them out. However, we'd much rather do this in middleware because (a) we would rather not write validation code for each prisma query, and (b) we use typegraphql with prisma, so the code that invokes prisma is auto-generated and hard to modify

**Reply by ryands17** [reply]
@mdesousa 👋 

As of now, I would then suggest keeping the field as optional because Middleware runs after the validation phase. It would be great if you could create a [feature request](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=feature_request.md&title=) for this specific use case so that we can look into this 🙂

**Comment by mdesousa** [comment, marked as answer]
Filed #15592, thanks.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
